### PR TITLE
Reomve on load dir

### PR DESCRIPTION
**Pull request by rgrinberg**
This is the last instance of on_load_dir. @diml should I get rid of this functionality in build_system.ml?

I could also memoize generating the `Path.Map.t` map for better perf.

### CODE REVIEW

File: src/local_package.ml [hunk 1]
```diff
@@ -10,8 +12,22 @@ type t =
   ; pkg : Package.t
   ; libs : Lib.Set.t
   ; virtual_lib : Lib.t option Lazy.t
+  ; id : Id.t
```
Comment (aalekseyev) [inline src/local_package.ml]: I think we should not have the id and remove the hash function and use physical equality instead.
The motivation is to no have multiple objects around with different contents that compare equal unnecessarily.
I pushed the commit that makes that change. Feel free to revert that if you disagree or if I'm missing the point.

File: src/local_package.ml [hunk 2]
```diff
@@ -142,3 +182,15 @@ let virtual_lib t = Lazy.force t.virtual_lib
 
 let meta_template t =
   Path.extend_basename (meta_file t) ~suffix:".template"
+
+let defined_in sctx ~dir =
+  let local_packages = of_sctx sctx in
+  let by_build_dir =
```
Comment (aalekseyev) [inline src/local_package.ml]: Should `by_build_dir` become a part of a memoized value?
I'm not sure how often this is called. Is it once for each directory?
If we have a big repo with many hundreds of directories and packages, that might add up?

**Comment by diml** [comment]
> @diml should I get rid of this functionality in build_system.ml?

Yes!

File: src/local_package.ml (hunk 2 above)
Comment (rgrinberg) [reply src/local_package.ml]: Yeah, I think that's good. Should we just memoize generating the `Path.Map.t` itself however?

We could also make accessing it a memoized function as well. That will give us another improvement for cases where the map changes but the key we are interested in does not.

**Comment by rgrinberg** [comment]
@aalekseyev have another look. We now need to use the output of `of_sctx` as input so we need a hashing function as well.

Shall I memoize the lookup in the map? It seems nice not to re-generate the rules for all packages if only 1 package is added (or removed) for example.

**Comment by rgrinberg** [comment]
I will write a better hashing function for `Local_package`. I'm thinking that something that takes the build dir and package names into account should be good enough.

**Comment by rgrinberg** [comment]
@aalekseyev, can you have a look at the last commit? Sorry, I know you approved, but I want to double check.

**Comment by aalekseyev** [comment]
Yeah, all those hash functions in the last commit look like functions, which is what counts. :-)

(the `Option` one encourages systematic collisions between `None` and `Some None`, which is unfortunate, but it seems unlikely to be much of a problem)

The commit before also looks good to me.

**Pull request merged**